### DISCUSSION TITLE
A Better Link

### DISCUSSION THREAD

**Discussion by machineghost** [vercel/next.js#23799] [Ideas]
### Describe the feature you'd like to request

Link is ... not the friendliest component, especially for library newcomers.  For one thing, all of its documentation explicitly shows that you need to nest (redundant) `<a>` tags inside of it ... but when you do that you get:

> Warning: validateDOMNesting(...): `<a>` cannot appear as a descendant of `<a>`.

But it's not just the documentation: there's also all the issues with having to wrap Links in fragments, forward refs, etc. ... just to make what _should_ be one of the simplest/most basic building blocks of any site.

And then there's styling: it's literally impossible to style the Link tag directly (it doesn't take a `style` prop), and if you want to use a library like StyledComponents instead there are more issues there.

### Describe the solution you'd like

Give Next developers a simpler alternative (Link2? EasyLink?).  I'm sure the current `Link` handles lots of complex cases that require it's current complexity ... but surely that complexity isn't always necessary for every `<a>` on every site?

The dream version of Link would not require you to even *think* about refs to use it.  It would not require any wrapping fragments/tags to use, nor would it require redundant `<a>` tags inside.  It would take a `style` prop, be styleable with StyledComponents ("out of the box") ... and just generally it would be a drop-in replacement for an `<a>`: it would be a brain dead alternative to the current "you need a masters in React/Next.js to use this component" Link ;)

### Describe alternatives you've considered

Making a 3rd party version of such a component ... but surely this belongs in the core framework itself?  And hopefully the wealth of other Link-related issues (and all the thumbs up "votes" inside them) demonstrate the potential value?

To put it another way, if you forgot all the "plumbing" of Next.js for a moment, and asked yourself "if I could just design the easiest/simplest/most welcoming version of this component possible, what would it look like?" ... and then built *that* component (using all your knowledge of Next.js plumbing) ... I think it would be a vastly superior component to the current version, which feels (to me at least) like it was built in the opposite direction.

**Comment by leodr** [comment]
I don't think the `<Link>` component is that bad. It is not a drop-in replacement for an `<a>`-tag, but it is pretty low effort:

1. Wrap the `<a>`-tag with `<Link>`
2. Take the `href`-attribute from the `<a>`-tag and stick it onto `<Link>`

I'm also unsure if it is worth at this point to make a breaking change to the `<Link>` since virtually **every** Next.js project uses it extensively.

**Reply by machineghost** [reply]
>Wrap the <a>-tag with <Link>
>Take the href-attribute from the <a>-tag and stick it onto <Link>

It's very much *not* that simple, and I specifically outlined three reasons why it's not in my post.

A) It's not simple because of all the ref stuff.  If you had anything inside your `<a>` you then need to wrap things in fragments, or pass forward refs, or all that jazz.

B) As a corollary of that,  it's also not because if you were using `StyledComponents` on that `<a>` tag: they won't convert cleanly. 

C) It's also not because if you have a style property on that `<a>` tag, it gets lost when you convert to `<Link>`.

But hey, it's not like anyone ever puts other tags inside their links, or ever tries to style their links: those are crazy unusual use cases ... right?

>I'm also unsure if it is worth at this point to make a breaking change to the <Link>

Perhaps I was unclear: I am *not* requesting a breaking change `<Link>`.  I am asking for a *new* tag (eg. `<SimpleLink>`, `<EasyLink>`, `<A>`, etc.) ... since clearly breaking Link is not viable.

**Reply by machineghost** [reply]
At the _absolute, most minimal_, there is a bug here in that the documentation tells you to put `<a>` tags inside `<link>` tags ... which then generates warnings.

**Reply by leodr** [reply]
> At the _absolute, most minimal_, there is a bug here in that the documentation tells you to put `<a>` tags inside `<link>` tags ... which then generates warnings.

I don't know what you're talking about. I've just created a fresh application with `create-next-app` and used this code to create a link:

```js
<Link href="/second">
    <a>To second</a>
</Link>
```

I did not get a warning and the DOM looks like this:

![image](https://user-images.githubusercontent.com/39763575/114044138-39a42e00-9887-11eb-9e0e-30f1a301eaf3.png)

Just as one would expect.

**Reply by leodr** [reply]
And could you provide an example of what you mean with the ref stuff? And what exactly is broken when you use styled-components with the `<a>`-tag?

Also, no it should not be common to put any other tag inside of the link. I think you should use the `<Link>` component like you would use `<a>` tags in regular HTML. For everything else use `onClick`-events and `router.push()`.

**Comment by machineghost** [comment]
When I have this code:

    <Link href={href}>
        <a style={isPublic ? null : { color: '#90bceb' }}>{text}</a>
    </Link>

I get:

> index.tsx?8abf:521 Warning: validateDOMNesting(...): <a> cannot appear as a descendant of <a>.
>     at a
>     at a
>     at Link (webpack-internal:///./node_modules/next/dist/client/link.js:88:19)
>     at Link (webpack-internal:///./src/components/ui/basic.js:159:24)
>     at li
>     at ul
>     at pages__RecommendationList (webpack-internal:///./node_modules/styled-components/dist/styled-components.browser.esm.js:28:19542)
>     at div
>     at pages__Content (webpack-internal:///./node_modules/styled-components/dist/styled-components.browser.esm.js:28:19542)
>     at div
>     at layout__InnerPage (webpack-internal:///./node_modules/styled-components/dist/styled-components.browser.esm.js:28:19542)
>     at div
>     at layout__FrontPage (webpack-internal:///./node_modules/styled-components/dist/styled-components.browser.esm.js:28:19542)
>     at main
>     at layout__Main (webpack-internal:///./node_modules/styled-components/dist/styled-components.browser.esm.js:28:19542)
>     at Ge (webpack-internal:///./node_modules/styled-components/dist/styled-components.browser.esm.js:28:17052)
>     at Layout (webpack-internal:///./src/components/layout.js:80:29)
>     at IndexPage (webpack-internal:///./pages/index.js:54:19)
>     at App (webpack-internal:///./pages/_app.js:26:24)
>     at ErrorBoundary (webpack-internal:///./node_modules/@next/react-dev-overlay/lib/internal/ErrorBoundary.js:23:47)
>     at ReactDevOverlay (webpack-internal:///./node_modules/@next/react-dev-overlay/lib/internal/ReactDevOverlay.js:73:23)
>     at Container (webpack-internal:///./node_modules/next/dist/client/index.js:149:5)
>     at AppContainer (webpack-internal:///./node_modules/next/dist/client/index.js:637:24)
>     at Root (webpack-internal:///./node_modules/next/dist/client/index.js:768:24)

Changing the code to:

    <Link href={href}>
        <span style={isPublic ? null : { color: '#90bceb' }}>{text}</span>
    </Link>

Removes the error.

**Reply by machineghost** [reply]
>And what exactly is broken when you use styled-components with the <a>-tag?

I apologize for this one; I'd previously had a problem with using StyledComponents and Links, but I just tested now and `styled(Link)` works without issue ... so there are only two things, not three, that make `<a>` => `<Link>` not trivial.

**Reply by leodr** [reply]
Cannot reproduce unfortunately. With the code you provided:

```js
<Link href="/second">
    <a style={isPublic ? null : { color: "#90bceb" }}>To second</a>
</Link>
```

No warning in the console, DOM looks like this (no nested `<a>`-tag):

![image](https://user-images.githubusercontent.com/39763575/114045897-acfa6f80-9888-11eb-9f89-044eaf7c6b78.png)

**Reply by machineghost** [reply]
Could I have an out of date version?  I'm on 10.0.6. I'll go install latest to double check, but could you please confirm what version you are on?

**Reply by machineghost** [reply]
Gah, nevermind!  I am sorry to have wasted your time, I found the issue and it was in my code :(

Thank you so much.  Please consider this closed.

**Reply by leodr** [reply]
Got version 10.1.3, React 17.0.2

**Reply by leodr** [reply]
> Gah, nevermind! I am sorry to have wasted your time, I found the issue and it was in my code :(
> 
> Thank you so much. Please consider this closed.

Got ya, no problem. I think you can somewhere mark the discussion as answered.

**Comment by machineghost** [comment]
I don't see a way to unfortunately, otherwise I would.

And in case you were curious, I forgot we had a replacement wrapper around Link ... made to workaround that `ref` issue :(

**Reply by leodr** [reply]
Oh ok, makes sense. I’m glad we found the issue👍

**Comment by machineghost** [comment]
One last thing: do you know of any "low hanging fruit" tickets that are "PR requested" ... that I might tackle to make up for your lost time?

**Reply by leodr** [reply]
Are you asking me? Well, I don't work for Vercel or Next.js, so it's all good.

If you're looking to contribute there are issues labelled with ["Good first issue"](https://github.com/vercel/next.js/labels/good%20first%20issue) that you can have a look at. Adding examples is usually a quick and easy task that helps out other developers.

**Reply by machineghost** [reply]
Oh, LOL I just assumed without even checking your title.  I'll check that label out, thanks.
